### ISSUE TITLE
Building with OpenJDK 11

### ISSUE THREAD

**Issue by df7cb** [tada/pljava#212]
OpenJDK 8 was removed from Debian/unstable, so the trick "build with Java 8, run with Java 11" no longer works.

Trying to build with openjdk-11-headless results in:
```
[INFO] ---------------------< org.postgresql:pljava-api >----------------------
[INFO] Building PL/Java API 1.5.3-SNAPSHOT                                [2/8]
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ pljava-api ---
[INFO] 
[INFO] --- maven-resources-plugin:3.0.1:resources (default-resources) @ pljava-api ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /home/cbe/projects/postgresql/pljava/pljava.git/pljava-api/src/main/resources
[INFO] 
[INFO] --- maven-compiler-plugin:2.5.1:compile (default-compile) @ pljava-api ---
[INFO] Compiling 24 source files to /home/cbe/projects/postgresql/pljava/pljava.git/pljava-api/target/classes
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] Failure executing javac, but could not parse the error:
Fatal Error: Unable to find package java.lang in classpath or bootclasspath

[INFO] 1 error
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for PostgreSQL PL/Java 1.5.3-SNAPSHOT:
[INFO] 
[INFO] PostgreSQL PL/Java ................................. SUCCESS [  4.198 s]
[INFO] PL/Java API ........................................ FAILURE [  1.567 s]
[INFO] PL/Java backend Java code .......................... SKIPPED
[INFO] PL/Java backend native code ........................ SKIPPED
[INFO] PL/Java Deploy ..................................... SKIPPED
[INFO] PL/Java Ant tasks .................................. SKIPPED
[INFO] PL/Java examples ................................... SKIPPED
[INFO] PL/Java packaging .................................. SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  6.443 s
[INFO] Finished at: 2019-04-01T09:24:29+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:2.5.1:compile (default-compile) on project pljava-api: Compilation failure
[ERROR] Failure executing javac, but could not parse the error:
[ERROR] Fatal Error: Unable to find package java.lang in classpath or bootclasspath
```

We'd need a new release that supports a newer OpenJDK for apt.postgresql.org. Thanks!

**Comment by jcflack** [comment]
Hi Christoph,

Thanks for the heads up. Is there a bit of lead time available on this? How long before sid becomes a version people are clamoring for?

Java 9 brought big changes that might require substantial edits in the build system, and I am not sure it will be practical to make those edits in the PL/Java 1.5.x branch while also keeping it buildable on older versions. So far, my (somewhat inchoate) plan has been to keep 1.5.x as the branch that everybody can still build on ancient systems, and make the-major-version-that-master-is-becoming be the modern one, that will probably build *only* on Java 9 and later. But I still have a lot of work planned for that branch.

Meanwhile, a little playing around this evening suggests that not-very-extensive changes to the 1.5.x pom will allow it to go further:

    --- a/pom.xml
    +++ b/pom.xml
    @@ -90,15 +90,12 @@
                                            <source>1.6</source>
                                            <target>1.6</target>
                                            <encoding>${project.build.sourceEncoding}</encoding>
    -                                       <compilerArguments>
    -                                               <bootclasspath>${env.JAVA_HOME}/jre/lib/rt.jar</bootclasspath>
    -                                       </compilerArguments>
                                    </configuration>
                            </plugin>
                            <plugin>
                                    <groupId>org.apache.maven.plugins</groupId>
                                    <artifactId>maven-jar-plugin</artifactId>
    -                               <version>2.6</version>
    +                               <version>3.0.2</version>
                            </plugin>
                            <plugin>
                                    <groupId>org.apache.maven.plugins</groupId>

Honestly, I don't know what that `bootclasspath` entry was ever doing there to begin with. Thomas added it back in 2013, and the commit message doesn't say why. Taking it out does not break the current build for me with Java 8 at all. If building with Java 12 instead of 11, the `source` and `target` should both be changed (to a minimum of 7, as 12 dropped support for building for targets 6 and earlier).

Perhaps as an interim solution, you could build with a patch to the pom?

However, the above isn't enough yet. With those changes, it gets as far as the embedded JavaScript and then fails. That seems to be related to [this Maven issue][MNG-6275], which looks like it has a complicated history, maybe it was merged in Maven 3.5.1, or 3.5.2, but has been reopened, and the last comment (as of now) says it was reverted. It is marked critical, so I assume there will be progress....

So far, I have not tried to simply download a more recent Maven to see how far that gets. What version are you building with?

[MNG-6275]: https://issues.apache.org/jira/browse/MNG-6275

**Comment by ChristophBerg** [comment]
Re: Chapman Flack 2019-04-02 <tada/pljava/issues/212/478834260@github.com>
> Thanks for the heads up. Is there a bit of lead time available on this? How long before sid becomes a version people are clamoring for?

Hi,

unfortunately I only noticed when our "all packages have satisfiable
build-deps" check went off the other day for Debian/unstable, and
today the removal of openjdk-8 has propagated to Debian/testing which
will eventually be released as Buster. For users, this is still not a
problem because no stable release is affected, and in the development
branches, the packages still work at run-time. We just couldn't
rebuild them if we had to.

If it takes time to fix this, I can mute that check for pljava, so it
isn't urgent, just annoying.

> Java 9 brought big changes that might require substantial edits in the build system, and I am not sure it will be practical to make those edits in the PL/Java 1.5.x branch while also keeping it buildable on older versions. So far, my (somewhat inchoate) plan has been to keep 1.5.x as the branch that everybody can still build on ancient systems, and make the-major-version-that-master-is-becoming be the modern one, that will probably build *only* on Java 9 and later. But I still have a lot of work planned for that branch.

Maintaining separate packages for different branches is possible in
theory, but practically I'd probably stop building packages for the
unsupported combinations; the packages won't get removed so people can
still use them.

Availability looks like this:

Debian:
openjdk-11 | 11.0.1+13-2~bpo9+1 | stretch-backports | source
openjdk-11 | 11.0.2+9-3~bpo9+1  | stretch-backports | source
openjdk-11 | 11.0.3+1-1~bpo9+1  | backports-policy  | source
openjdk-11 | 11.0.3+1-1         | testing           | source
openjdk-11 | 11.0.3+4-3         | buildd-unstable   | source
openjdk-11 | 11.0.3+4-3         | unstable          | source

openjdk-12 | 12.0.0-3      | unstable   | source

Ubuntu:
openjdk-8 | 8u77-b03-3ubuntu3          | xenial                   | source
openjdk-8 | 8u162-b12-1                | bionic/universe          | source
openjdk-8 | 8u181-b13-1                | cosmic/universe          | source
openjdk-8 | 8u191-b12-2ubuntu0.16.04.1 | xenial-security          | source
openjdk-8 | 8u191-b12-2ubuntu0.16.04.1 | xenial-updates           | source
openjdk-8 | 8u191-b12-2ubuntu0.18.04.1 | bionic-security/universe | source
openjdk-8 | 8u191-b12-2ubuntu0.18.04.1 | bionic-updates/universe  | source
openjdk-8 | 8u191-b12-2ubuntu0.18.10.1 | cosmic-security/universe | source
openjdk-8 | 8u191-b12-2ubuntu0.18.10.1 | cosmic-updates/universe  | source
openjdk-8 | 8u212-b01-1                | disco/universe           | source

openjdk-9 | 9~b114-0ubuntu1 | xenial/universe | source

openjdk-12 | 12.0.0-3 | disco/universe | source

Apparently bionic (LTS) and cosmic (non-LTS) have openjdk-8 only.

> Perhaps as an interim solution, you could build with a patch to the pom?
> 
> However, the above isn't enough yet.

I could, but it if still fails...

> So far, I have not tried to simply download a more recent Maven to see how far that gets. What version are you building with?

maven      | 3.3.9-4       | stable     | source, all
maven      | 3.6.0-1       | testing    | source, all
maven      | 3.6.0-1       | unstable   | source, all

Christoph

**Comment by jcflack** [comment]
No joy ... tried with Maven 3.5.2 and 3.6.0, still falls over right at the first javascript bit.

I've added a comment on the [MNG-6275][] issue asking if that's where the trouble is, and the outlook for a resolution if so.

[MNG-6275]: https://issues.apache.org/jira/browse/MNG-6275

**Comment by jcflack** [comment]
Another thing worth knowing:

Today saw the first production release of [GraalVM, 19.0][graalvm].

This release of GraalVM is (among other things) a Java SE 8 JVM, and PL/Java 1.5.2 needs no special attention to build and/or run with it (other than to add `-Dpolyglot.js.nashorn-compat=true` on the `mvn` command line).

It would be great to see GraalVM packaged for the distro, as it can do very cool things (as can PL/Java, when running on it).

[graalvm]: https://github.com/oracle/graal/releases/tag/vm-19.0.0

**Comment by jcflack** [comment]
I've pushed a branch `trackjdk/REL1_5_STABLE/java9` that will build for me in the post-Java-8 JDKs (but will also still build with the older ones), if you have time to give it a try.

Java 12 eliminated the ability to build with runtime compatibility back to 6. If built with 12, the classfiles will only be recognized by Java 7 and later JREs.

This branch is based on the latest merges to `REL1_5_STABLE`, so it should also work with PG 12. If it works and I merge it, a release supporting PG 12 and building with later JREs could shortly follow.

**Comment by df7cb** [comment]
Hi,
I finally got to test this. It's complaining it can't find `jni.h`:
```
export JAVA_HOME = /usr/lib/jvm/default-java
mvn --batch-mode -Pwnosign -Psaxon-examples -Dso.debug=true -Dpgsql.pgconfig=/usr/lib/postgresql/11/bin/pg_config -Dpljava.libjvmdefault=/usr/lib/jvm/java-11-openjdk-amd64/lib/server/libjvm.so clean install
...
[ERROR] /home/cbe/projects/postgresql/pljava/pljava.git/pljava-so/src/main/include/pljava/pljava.h:16:10: fatal error: jni.h: Datei oder Verzeichnis nicht gefunden
...
```
The file is there, though:
```
openjdk-11-jdk-headless:amd64: /usr/lib/jvm/java-11-openjdk-amd64/include/jni.h
```

**Comment by jcflack** [comment]
On 6/24/19 11:22 AM, Christoph Berg wrote:

> [ERROR] /home/cbe/projects/postgresql/pljava/pljava.git/pljava-so/src/main/include/pljava/pljava.h:16:10: fatal error: jni.h: Datei oder Verzeichnis nicht gefunden

That's weird. It certainly did not happen in my environment.

I am not there at the moment. Would it be possible to run mvn with -X
and attach the output (either all of it, or trimmed down to the command-
line arguments passed to gcc)?

**Comment by jcflack** [comment]
In my environment it is finding a copy of `jni.h` that was bundled in the `gcc` package, of all places.

I had never noticed that. That is not very reassuring.

**Comment by jcflack** [comment]
I think the current tip of the `trackjdk/REL1_5_STABLE/java9` branch should work now.

**Comment by df7cb** [comment]
Confirmed with Java 11 and PG11 and 12 on Debian unstable.
On the other side of the spectrum, there's a problem though. Xenial, Java 8, PG9.5:
```
2019-06-26 14:32:09.362 CEST [3580] cbe@postgres FEHLER:  java.sql.SQLException: SPI function SPI_prepare failed with error SPI_ERROR_UNCONNECTED
2019-06-26 14:32:09.362 CEST [3580] cbe@postgres ANWEISUNG:  CREATE EXTENSION pljava
```

**Comment by jcflack** [comment]
This patch isn't what broke that, so I've opened issue #221 for that (`SPI_ERROR_UNCONNECTED` in PG 9.6 and earlier). I'll go ahead and pull-request this patch, and then deal with #221 shortly.

**Comment by df7cb** [comment]
Let me know when there's a branch that you'd like having tested!

**Comment by jcflack** [comment]
A branch worth testing now would be `REL1_5_STABLE`. :)

**Comment by df7cb** [comment]
I finally tested 1_5 and it looks good. Are you planning to do a release soonish?

**Comment by jcflack** [comment]
Thanks for checking. Was it a recent `REL1_5_STABLE`? There were a couple pull requests merged just last night.

At the moment, I am holding "soonish" to be "some time between this month and general availability of PostgreSQL 12". I still have WIP on #225 that will probably go in. I continue to be of two minds about #200 for this release ... most of the identified bugs are fixed (in `bug/REL1_5_STABLE/issue200`) but some of the fixes would change old behavior, and the more I learn about the old behavior, the longer the #200 bug report gets.

Also there's been a lot of progress completing the Saxon example to fill in the ISO SQL XQuery-based functionality that isn't in PG core. That's on a WIP branch here that hasn't been pushed to github yet but probably will after another rebase or two, and be part of the release.

**Comment by df7cb** [comment]
It was from today, c9e48ed341f53837.

Maybe don't wait for everything to be perfect? You could still do another release once 12 is out.

**Comment by jcflack** [comment]
Ok, I'll say you've talked me into saving the #200 fixes for a later release. #225 is in, and the fleshed-out XQuery example. `REL1_5_STABLE` right now should be pretty much it. I have just committed release notes.

**Comment by jcflack** [comment]
Believed resolved in 1.5.3.

**Comment by df7cb** [comment]
Hi,
sorry for not having tested that yet. I've been busy with the PG12 release and other bits. Will give it a try next week.
Thanks!

**Comment by df7cb** [comment]
I'm afraid it doesn't work with java 11:
```
+ mvn --batch-mode -Pwnosign -Psaxon-examples -Dso.debug=true -Dpgsql.pgconfig=/usr/lib/postgresql/11/bin/pg_config -Dpljava.libjvmdefault=/usr/lib/jvm/java-11-openjdk-amd64/lib/server/libjvm.so clean install
[...]
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 74 source files to /home/cbe/projects/postgresql/pljava/postgresql-pljava/pljava/target/classes
[INFO] /home/cbe/projects/postgresql/pljava/postgresql-pljava/pljava/src/main/java/org/postgresql/pljava/internal/Session.java: /home/cbe/projects/postgresql/pljava/postgresql-pljava/pljava/src/main/java/org/postgresql/pljava/internal/Session.java uses or overrides a deprecated API.
[INFO] /home/cbe/projects/postgresql/pljava/postgresql-pljava/pljava/src/main/java/org/postgresql/pljava/internal/Session.java: Recompile with -Xlint:deprecation for details.
[INFO] /home/cbe/projects/postgresql/pljava/postgresql-pljava/pljava/src/main/java/org/postgresql/pljava/jdbc/AbstractResultSet.java: Some input files use unchecked or unsafe operations.
[INFO] /home/cbe/projects/postgresql/pljava/postgresql-pljava/pljava/src/main/java/org/postgresql/pljava/jdbc/AbstractResultSet.java: Recompile with -Xlint:unchecked for details.
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /home/cbe/projects/postgresql/pljava/postgresql-pljava/pljava/src/main/java/org/postgresql/pljava/jdbc/SQLXMLImpl.java:[597,54] cannot find symbol
  symbol:   method newFactory()
  location: class javax.xml.stream.XMLInputFactory
[ERROR] /home/cbe/projects/postgresql/pljava/postgresql-pljava/pljava/src/main/java/org/postgresql/pljava/jdbc/SQLXMLImpl.java:[784,78] cannot find symbol
  symbol:   method newFactory()
  location: class javax.xml.stream.XMLInputFactory
[ERROR] /home/cbe/projects/postgresql/pljava/postgresql-pljava/pljava/src/main/java/org/postgresql/pljava/jdbc/SQLXMLImpl.java:[1121,80] cannot find symbol
  symbol:   method newFactory()
  location: class javax.xml.stream.XMLOutputFactory
[ERROR] /home/cbe/projects/postgresql/pljava/postgresql-pljava/pljava/src/main/java/org/postgresql/pljava/jdbc/SQLXMLImpl.java:[2899,71] cannot find symbol
  symbol:   method newFactory()
  location: class javax.xml.stream.XMLInputFactory
[ERROR] /home/cbe/projects/postgresql/pljava/postgresql-pljava/pljava/src/main/java/org/postgresql/pljava/jdbc/SQLXMLImpl.java:[2901,72] cannot find symbol
  symbol:   method newFactory()
  location: class javax.xml.stream.XMLOutputFactory
[INFO] 5 errors 
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for PostgreSQL PL/Java 1.5.3:
[INFO] 
[INFO] PostgreSQL PL/Java ................................. SUCCESS [  6.721 s]
[INFO] PL/Java API ........................................ SUCCESS [  5.011 s]
[INFO] PL/Java backend Java code .......................... FAILURE [  1.324 s]
[INFO] PL/Java backend native code ........................ SKIPPED
[INFO] PL/Java Deploy ..................................... SKIPPED
[INFO] PL/Java Ant tasks .................................. SKIPPED
[INFO] PL/Java examples ................................... SKIPPED
[INFO] PL/Java packaging .................................. SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  13.610 s
[INFO] Finished at: 2019-10-10T11:44:50+02:00
```
```
$ java -version
openjdk version "11.0.5-ea" 2019-10-15
OpenJDK Runtime Environment (build 11.0.5-ea+9-post-Debian-1)
OpenJDK 64-Bit Server VM (build 11.0.5-ea+9-post-Debian-1, mixed mode, sharing)
```

**Comment by jcflack** [comment]
Oh bother. I didn't actually test in Java 6 this time, or in 11, I tested in 7, 8, and 12.

Those methods were introduced in 7. The pre-java9 compilers did not catch the problem because the `-source 1.6`/`-target 1.6` options only affect syntax and code generation, but don't verify that the API is limited to the named release. (The generated code would have thrown a runtime `LinkageError` at those points if run on a 1.6 VM.)

The smarter `--release 6` option used with the newer compilers actually checks that all the API existed in the chosen release. My testing with Java 12 wasn't sufficient to cover 11, because 12 drops support for release 6.

Looks like I'll achieve a shorter interval between 1.5.3 and 1.5.4 than there was from .1 to .2.

**Comment by jcflack** [comment]
Continued in #235.

**Comment by jcflack** [comment]
Hi Christoph,

If there's any chance you could test `REL1_5_STABLE` before I make it another release, it would be great to know if you experience any problem again.

Regards,
-Chap

**Comment by df7cb** [comment]
I just tested on sid (Java 11) and on xenial (Java 8), for PG 9.4 .. 12. Everything fine.

Let me know if there's any interesting distributions with Java versions in between that I should cover as well.

**Comment by jcflack** [comment]
Thanks! I guess zesty and artful had Java 9 packages, but I've built on AdoptOpenJDK 9 and 10 here, so I think I'll take the chance. I've released 1.5.4.
